File: projects/gnomad-api/src/schema/datasets/gnomad_r2_1/gnomadV2Constraint.js
import { GraphQLFloat, GraphQLInt, GraphQLList, GraphQLObjectType, GraphQLString } from 'graphql'

import { UserVisibleError } from '../../errors'

export const GnomadConstraintType = new GraphQLObjectType({
  name: 'GnomadConstraint',
  fields: {
    // Expected
    exp_lof: { type: GraphQLFloat },
    exp_mis: { type: GraphQLFloat },
    exp_syn: { type: GraphQLFloat },
    // Observed
    obs_lof: { type: GraphQLInt },
    obs_mis: { type: GraphQLInt },
    obs_syn: { type: GraphQLInt },
    // Observed/Expected
    oe_lof: { type: GraphQLFloat },
    oe_lof_lower: { type: GraphQLFloat },
    oe_lof_upper: { type: GraphQLFloat },
    oe_mis: { type: GraphQLFloat },
    oe_mis_lower: { type: GraphQLFloat },
    oe_mis_upper: { type: GraphQLFloat },
    oe_syn: { type: GraphQLFloat },
    oe_syn_lower: { type: GraphQLFloat },
    oe_syn_upper: { type: GraphQLFloat },
    // Z
    lof_z: { type: GraphQLFloat },
    mis_z: { type: GraphQLFloat },
    syn_z: { type: GraphQLFloat },
    // Other
    pLI: { type: GraphQLFloat },
    flags: { type: new GraphQLList(GraphQLString) },
  },
})

export const fetchGnomadConstraintByTranscript = async (ctx, transcriptId) => {
  try {
    const response = await ctx.database.elastic.get({
      index: 'gnomad_2_1_1_constraint',
      type: 'documents',
      id: transcriptId,
    })

    return response._source
  } catch (err) {
    if (err.message === 'Not Found') {
      throw new UserVisibleError(`Constraint not found for transcript ${transcriptId}`)
    }
    throw err
  }
}

File: data-pipeline/src/data_pipeline/constraint/gnomad_v2_constraint.py
import hail as hl


def prepare_gnomad_v2_constraint(path):
    ds = hl.read_table(path)

    # Don't need the information in globals for the browser
    ds = ds.select_globals()

    # Select relevant fields
    ds = ds.select(
        # ID
        transcript_id=ds.transcript,
        gene_id=ds.gene_id,
        # Expected
        exp_lof=ds.exp_lof,
        exp_mis=ds.exp_mis,
        exp_syn=ds.exp_syn,
        # Observed
        obs_lof=ds.obs_lof,
        obs_mis=ds.obs_mis,
        obs_syn=ds.obs_syn,
        # Observed/Expected
        oe_lof=ds.oe_lof,
        oe_lof_lower=ds.oe_lof_lower,
        oe_lof_upper=ds.oe_lof_upper,
        oe_mis=ds.oe_mis,
        oe_mis_lower=ds.oe_mis_lower,
        oe_mis_upper=ds.oe_mis_upper,
        oe_syn=ds.oe_syn,
        oe_syn_lower=ds.oe_syn_lower,
        oe_syn_upper=ds.oe_syn_upper,
        # Z
        lof_z=ds.lof_z,
        mis_z=ds.mis_z,
        syn_z=ds.syn_z,
        # Other
        pli=ds.pLI,
        flags=ds.constraint_flag,
    )

    ds = ds.key_by("transcript_id")

    return ds
